### ISSUE TITLE
I can not install node-rdkafka in node-v6.11.0

### ISSUE THREAD

**Issue by thekingofworld** [Blizzard/node-rdkafka#216]
env: 
    node version : v6.11.0
    npm  version : v3.10.10

npm-error-log:
[npm-debug.txt](https://github.com/Blizzard/node-rdkafka/files/1126455/npm-debug.txt)

**Comment by thekingofworld** [comment]
I have run `export WITH_SASL=0` before `npm install`, but also not succeed

**Comment by webmakersteve** [comment]
Can you force install from npm, `cd` into the directory, and then try to manually run `node-gyp rebuild` and give me the output from that? Unfortunately NPM does not report much more than the status code of gyp

**Comment by thekingofworld** [comment]
the follows is output, thanks
----------------------------------
gyp info it worked if it ends with ok
gyp info using node-gyp@3.6.2
gyp info using node@6.11.0 | linux | x64
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp info spawn /usr/local/bin/python2
gyp info spawn args [ '/down/node-v6.11.0-linux-x64/lib/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/opt/htdocs/liyao/node_practice/rdkafka/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/down/node-v6.11.0-linux-x64/lib/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/root/.node-gyp/6.11.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/root/.node-gyp/6.11.0',
gyp info spawn args   '-Dnode_gyp_dir=/down/node-v6.11.0-linux-x64/lib/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/root/.node-gyp/6.11.0/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/opt/htdocs/liyao/node_practice/rdkafka',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
gyp: binding.gyp not found (cwd: /opt/htdocs/liyao/node_practice/rdkafka) while trying to load binding.gyp
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/down/node-v6.11.0-linux-x64/lib/node_modules/node-gyp/lib/configure.js:336:16)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
gyp ERR! System Linux 2.6.32-642.1.1.el6.x86_64
gyp ERR! command "/down/node-v6.11.0-linux-x64/bin/node" "/down/node-v6.11.0-linux-x64/bin/node-gyp" "rebuild"
gyp ERR! cwd /opt/htdocs/liyao/node_practice/rdkafka
gyp ERR! node -v v6.11.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
--------------------------

**Comment by webmakersteve** [comment]
Unfortunately that didn't help much either :(

Is it getting to the point where it is compiling any files? Or does it fail before that? It looks like the configure step is failing in gyp which generally means you are missing a dependency.  

What operating system are you running on? Can you try to install sasl anyway to see if that makes it work?

**Comment by chanakyamandava** [comment]
I am running into same error, I am using windows 7 64 bit

Traceback (most recent call last):
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\gyp_main.py", line 16, in <module>
    sys.exit(gyp.script_main())
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 545, in script_main
    return main(sys.argv[1:])
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 538, in main
    return gyp_main(args)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 514, in gyp_main
    options.duplicate_basename_check)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 130, in Load
    params['parallel'], params['root_targets'])
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\input.py", line 2775, in Load
    variables, includes, depth, check, True)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\input.py", line 417, in LoadTargetBuildFile
    build_file_data, PHASE_EARLY, variables, build_file_path)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\input.py", line 1292, in ProcessVariablesAndConditionsInDict
    build_file)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\input.py", line 1307, in ProcessVariablesAndConditionsInList
    ProcessVariablesAndConditionsInDict(item, phase, variables, build_file)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\input.py", line 1266, in ProcessVariablesAndConditionsInDict
    ProcessConditionsInDict(the_dict, phase, variables, build_file)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\input.py", line 1139, in ProcessConditionsInDict
    build_file)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\input.py", line 1063, in EvalCondition
    cond_expr, true_dict, false_dict, phase, variables, build_file)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\input.py", line 1097, in EvalSingleCondition
    raise syntax_error
SyntaxError: invalid syntax while evaluating condition '${BUILD_LIBRDKAFKA:-1}==1' in binding.gyp at character 1.
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\configure.js:305:16)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:219:12)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebui
ld"
gyp ERR! cwd C:\Users\cmandava\nodered-rdkafka\node_modules\node-rdkafka
gyp ERR! node -v v6.11.3
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok
npm ERR! Windows_NT 6.1.7601
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install"
npm ERR! node v6.11.3
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE

npm ERR! node-rdkafka@1.0.6 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the node-rdkafka@1.0.6 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the node-rdkafka package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs node-rdkafka
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls node-rdkafka
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\cmandava\nodered-rdkafka\npm-debug.log

**Comment by chenfanggm** [comment]
same issue when install node-rdkafka@2.1.1 on CentOS with node v7.10.1, npm v4.2.0

**Comment by erik-stephens** [comment]
> Please run configure before using this script

Need to run `npm run configure` or its equivalent.  In the 2.1.1 version, `WITH_SASL` has no effect.  Would be nice to know how to have node-rdkafka configure librdkafka to do things like `--disable-sasl` now.

**Comment by chenfanggm** [comment]
@erik-stephens 
Even thought `npm run configure` may solve the directly run `node-gyp rebuild` after a npm install --force, the original problem of successfully install from `npm install node-rdkafka` still not get fixed. Thanks.

**Comment by webmakersteve** [comment]
@erik-stephens The WITH_SASL workaround was added for machines that didn't have the library available. Now that it sources the existence of the library before adding it as a dependency, I didn't think it was necessary anymore.

Do you still want to compile node-rdkafka without SASL when SASL is installed on the machine?

@chenfanggm Are you running npm install or just node-gyp rebuild? If you're going to use gyp manually you need to call the hooks - it is not meant to be self contained with gyp alone.

**Comment by erik-stephens** [comment]
@webmakersteve I was able to install & run on alpine with 2.0.0.  I've been familiarizing myself with node-gyp and librdkafka build options the past couple of days which is why I chimed in here.  Digging through the bits, I found it strange that there is no way to pass options to librdkafka's `configure` since I wasn't seeing its auto-detection working, but I was incorrectly reading the build output.  After re-visiting just now, I was able to get node-rdkafka installed from node:6-alpine base image but I had to hack up [librdkafka.gyp](https://github.com/Blizzard/node-rdkafka/blob/master/deps/librdkafka.gyp#L250) to not link libcrypto or libssl.  Fair to consider those dependencies burdensome enough to make optional?

To the original posters, apologies if I butted in.  Maybe try the `--unsafe` flag to your npm install command.
